File: apps/badminton-coach/analyze/badminton_coach/strokes.py
"""Find swings in a pose sequence and say what kind of shot each one was.

Mirrors ``web/js/core/strokes.js``. The reasoning, in short:

* The signal is racket-wrist speed *relative to the hips*, in trunk lengths per
  second. Hip-relative so that running to the back of the court is not read as a
  swing; in trunk lengths so one threshold covers a junior and an adult, and a
  phone 5 m away and one 15 m away.
* A stroke is a local maximum of that speed. Peak wrist speed is the standard
  stand-in for contact in racket-sports analysis: true contact is within a frame
  or two of it, and unlike the shuttle, the wrist is something a pose model can
  actually see.
* At contact the hand is expressed in the torso frame. Its height relative to the
  shoulders separates overhead from drive from underarm; its lateral position,
  signed so positive is always the racket side, separates forehand from backhand
  -- a backhand being exactly the shot where the racket hand has crossed the
  body's midline.
"""

from __future__ import annotations

import math
from dataclasses import dataclass, field

from . import landmarks as L
from .biomech import core_visibility
from .geometry import distance, midpoint, sub

NAN = float("nan")


@dataclass
class StrokeOptions:
    """Tunables for :func:`detect_strokes`. Speeds are in trunk lengths per second."""

    peak_speed: float = 6.0
    """A peak must reach this to count as a swing rather than a fidget."""

    quiet_speed: float = 2.5
    """Speed below which the arm counts as quiet, used to find a swing's edges."""

    refractory: float = 0.30
    """Minimum gap between two contacts; badminton exchanges rarely beat this."""

    max_backswing: float = 0.80
    max_follow_through: float = 0.60
    min_visibility: float = 0.5
    peak_window: int = 3
    """Half-width, in frames, of the neighbourhood a peak must dominate."""


@dataclass
class ShotThresholds:
    """Family boundaries, in trunk lengths from the shoulder line."""

    overhead_height: float = 0.10
    underarm_height: float = -0.45
    forehand_lateral: float = 0.28
    backhand_lateral: float = 0.02


def wrist_speed_series(frames, racket_arm: str, trunk_lengths) -> list[float]:
    """Racket-wrist speed relative to the hips, in trunk lengths per second."""
    wrist_index = L.side("WRIST", racket_arm)
    relative = [
        sub(f["world"][wrist_index], midpoint(f["world"][L.LEFT_HIP], f["world"][L.RIGHT_HIP]))
        for f in frames
    ]
    speed = [NAN] * len(frames)
    for i in range(1, len(frames) - 1):
        dt = frames[i + 1]["t"] - frames[i - 1]["t"]
        if dt <= 0:
            continue
        trunk = trunk_lengths[i]
        if not (trunk and math.isfinite(trunk) and trunk > 1e-6):
            continue
        speed[i] = distance(relative[i + 1], relative[i - 1]) / dt / trunk
    if len(speed) > 2:
        speed[0] = speed[1]
        speed[-1] = speed[-2]
    return speed


def classify_shot(hand: dict, thresholds: ShotThresholds | None = None) -> dict:
    """Name the shot from the hand's position in the torso frame at contact."""
    th = thresholds or ShotThresholds()
    height, lateral = hand["height"], hand["lateral"]

    if height >= th.overhead_height:
        height_class = "overhead"
    elif height >= th.underarm_height:
        height_class = "drive"
    else:
        height_class = "underarm"

    if lateral >= th.forehand_lateral:
        side_class = "forehand"
    elif lateral <= th.backhand_lateral:
        side_class = "backhand"
    else:
        side_class = "roundhead"

    # A shot played over the head is only its own category when it is actually
    # overhead; lower down, the same lateral band is a straight-on forehand.
    if side_class == "roundhead" and height_class != "overhead":
        side_class = "forehand"

    # Distance from the nearest decision boundary, as a rough confidence: shots
    # near a boundary are the ones a human would hesitate over too.
    side_margin = min(abs(lateral - th.forehand_lateral), abs(lateral - th.backhand_lateral))
    height_margin = min(abs(height - th.overhead_height), abs(height - th.underarm_height))
    confidence = max(0.0, min(1.0, min(side_margin, height_margin) / 0.25))

    return {
        "shot": f"{side_class}-{height_class}",
        "side": side_class,
        "height": height_class,
        "confidence": confidence,
    }


def _local_peaks(speed, options: StrokeOptions) -> list[int]:
    w = options.peak_window
    peaks = []
    for i in range(w, len(speed) - w):
        v = speed[i]
        if not (v is not None and math.isfinite(v) and v >= options.peak_speed):
            continue
        dominant = True
        for j in range(i - w, i + w + 1):
            if j == i:
                continue
            other = speed[j]
            if other is None or not math.isfinite(other):
                continue
            # `>=` on the left and `>` on the right keeps exactly one index on a plateau.
            if (other >= v) if j < i else (other > v):
                dominant = False
                break
        if dominant:
            peaks.append(i)
    return peaks


def _apply_refractory(peaks, speed, times, refractory: float) -> list[int]:
    # Strongest first, so that when two candidates are too close we keep the real
    # contact rather than whichever came first in time.
    kept: list[int] = []
    for p in sorted(peaks, key=lambda i: speed[i], reverse=True):
        if all(abs(times[p] - times[k]) >= refractory for k in kept):
            kept.append(p)
    return sorted(kept)


def _walk_to(speed, times, start: int, direction: int, limit: float, quiet: float) -> int:
    i = start
    while True:
        nxt = i + direction
        if nxt < 0 or nxt >= len(speed):
            break
        if abs(times[nxt] - times[start]) > limit:
            break
        i = nxt
        v = speed[i]
        if v is not None and math.isfinite(v) and v <= quiet:
            break
    return i


def _min_by(items, fn) -> float:
    values = [fn(x) for x in items]
    values = [v for v in values if v is not None and math.isfinite(v)]
    return min(values) if values else NAN


def _max_by(items, fn) -> float:
    values = [fn(x) for x in items]
    values = [v for v in values if v is not None and math.isfinite(v)]
    return max(values) if values else NAN


def detect_strokes(frames, options: StrokeOptions | None = None,
                   thresholds: ShotThresholds | None = None) -> list[dict]:
    """Detect every stroke in a sequence of analysed frames.

    Each frame is ``{"t", "world", "metrics", "image"?}`` with ``world`` already
    in the y-up frame and ``metrics`` from :func:`biomech.frame_metrics`.
    """
    opts = options or StrokeOptions()
    th = thresholds or ShotThresholds()
    if len(frames) < 5:
        return []

    times = [f["t"] for f in frames]
    trunks = [f["metrics"]["trunk_length"] for f in frames]
    racket_arm = frames[0]["metrics"]["racket_arm"]
    speed = wrist_speed_series(frames, racket_arm, trunks)
    peaks = _apply_refractory(_local_peaks(speed, opts), speed, times, opts.refractory)

    strokes = []
    for peak in peaks:
        image = frames[peak].get("image")
        visibility = core_visibility(image) if image else 1.0
        if visibility < opts.min_visibility:
            continue

        start = _walk_to(speed, times, peak, -1, opts.max_backswing, opts.quiet_speed)
        end = _walk_to(speed, times, peak, +1, opts.max_follow_through, opts.quiet_speed)
        contact = frames[peak]["metrics"]
        label = classify_shot(contact["hand"], th)
        window = frames[start:end + 1]
        backswing = frames[start:peak + 1]

        strokes.append({
            "index": len(strokes),
            "frame": frames[peak].get("frame"),
            "t": times[peak],
            "start_t": times[start],
            "end_t": times[end],
            "start_frame": frames[start].get("frame"),
            "end_frame": frames[end].get("frame"),
            **label,
            "visibility": visibility,
            "peak_speed": speed[peak],
            "peak_speed_ms": speed[peak] * contact["trunk_length"],
            "backswing_duration": times[peak] - times[start],
            "follow_through_duration": times[end] - times[peak],
            "contact": {
                "elbow": contact["elbow"],
                "elbow_off": contact["elbow_off"],
                "shoulder_elevation": contact["shoulder_elevation"],
                "shoulder_elevation_off": contact["shoulder_elevation_off"],
                "shoulder_azimuth": contact["shoulder_azimuth"],
                "separation": contact["separation"],
                "trunk_lean": {k: v for k, v in contact["trunk_lean"].items() if k != "heading"},
                "knee_left": contact["knee_left"],
                "knee_right": contact["knee_right"],
                "stance_width": contact["stance_width"],
                "hand": {k: contact["hand"][k] for k in ("height", "lateral", "forward", "extension")},
            },
            # A coach reads the backswing as much as the contact: the deepest
            # elbow bend and the largest twist before contact are what the arm
            # had available to release.
            "backswing": {
                "min_elbow": _min_by(backswing, lambda f: f["metrics"]["elbow"]),
                "max_separation": _max_by(backswing, lambda f: f["metrics"]["separation"]),
                "max_shoulder_elevation": _max_by(backswing, lambda f: f["metrics"]["shoulder_elevation"]),
                "max_off_arm_elevation": _max_by(backswing, lambda f: f["metrics"]["shoulder_elevation_off"]),
            },
            "window": {
                "max_elbow": _max_by(window, lambda f: f["metrics"]["elbow"]),
                "min_knee": _min_by(window, lambda f: min(f["metrics"]["knee_left"], f["metrics"]["knee_right"])),
            },
        })
    return strokes


def summarise_strokes(strokes) -> dict:
    """Tally shots by name, for the session summary."""
    by_shot: dict[str, int] = {}
    for s in strokes:
        by_shot[s["shot"]] = by_shot.get(s["shot"], 0) + 1
    speeds = [s["peak_speed_ms"] for s in strokes if math.isfinite(s["peak_speed_ms"])]
    return {
        "count": len(strokes),
        "by_shot": by_shot,
        "forehand": sum(1 for s in strokes if s["side"] == "forehand"),
        "backhand": sum(1 for s in strokes if s["side"] == "backhand"),
        "roundhead": sum(1 for s in strokes if s["side"] == "roundhead"),
        "overhead": sum(1 for s in strokes if s["height"] == "overhead"),
        "mean_peak_speed_ms": sum(speeds) / len(speeds) if speeds else NAN,
        "max_peak_speed_ms": max(speeds) if speeds else NAN,
    }

File: apps/badminton-coach/web/js/core/strokes.js
/**
 * Find swings in a pose sequence and say what kind of shot each one was.
 *
 * ## Detecting the swing
 *
 * The signal is the speed of the racket wrist *relative to the hips*, in trunk
 * lengths per second. Two choices matter:
 *
 * - Relative to the hips, so that running to the back of the court does not read
 *   as a swing. MediaPipe's world landmarks are already hip-centred, so this
 *   falls out for free.
 * - In trunk lengths rather than metres, so one threshold works for a junior and
 *   an adult, and for a phone 5 m away and one 15 m away.
 *
 * A stroke is a local maximum of that speed. Peak wrist speed is a standard
 * stand-in for the moment of contact in racket-sports analysis: the true contact
 * is within a frame or two of it, and unlike the shuttle, the wrist is something
 * a pose model can actually see.
 *
 * ## Naming the shot
 *
 * At contact the hand is expressed in the torso frame (see `biomech.js`). Its
 * height relative to the shoulder line separates overhead shots from drives from
 * underarm lifts; its lateral position -- signed so that positive is always the
 * racket side -- separates forehand from backhand, because a backhand is
 * precisely the shot where the racket hand has crossed the body's midline.
 */

import { distance, sub } from './vec3.js';
import { LM, CORE, side } from './landmarks.js';
import { midpoint } from './vec3.js';
import { coreVisibility } from './biomech.js';

/** Tunables for {@link detectStrokes}. Speeds are in trunk lengths per second. */
export const DEFAULT_STROKE_OPTIONS = {
  /** A peak must reach this to count as a swing rather than a fidget. */
  peakSpeed: 6.0,
  /** Speed below which the arm counts as quiet, used to find the swing's edges. */
  quietSpeed: 2.5,
  /** Minimum gap between two contacts; badminton exchanges rarely beat this. */
  refractory: 0.30,
  /** How far back from contact to look for the start of the backswing. */
  maxBackswing: 0.80,
  /** How far forward from contact to look for the end of the follow-through. */
  maxFollowThrough: 0.60,
  /** Skip swings where the model could barely see the joints we are measuring. */
  minVisibility: 0.5,
  /** Half-width, in frames, of the neighbourhood a peak must dominate. */
  peakWindow: 3,
};

/** Thresholds separating shot families, in trunk lengths from the shoulder line. */
export const SHOT_THRESHOLDS = {
  overheadHeight: 0.10,
  underarmHeight: -0.45,
  forehandLateral: 0.28,
  backhandLateral: 0.02,
};

/**
 * Racket-wrist speed relative to the hips, per frame, in trunk lengths per second.
 *
 * @param {Array} samples frames of `{ t, world }` with world already y-up
 * @param {'left'|'right'} racketArm
 * @param {number[]} trunkLengths stable trunk length per frame, in metres
 */
export function wristSpeedSeries(samples, racketArm, trunkLengths) {
  const wristIndex = side('WRIST', racketArm);
  const relative = samples.map((s) => {
    const hip = midpoint(s.world[LM.LEFT_HIP], s.world[LM.RIGHT_HIP]);
    return sub(s.world[wristIndex], hip);
  });
  const speed = new Array(samples.length).fill(NaN);
  for (let i = 1; i < samples.length - 1; i += 1) {
    const dt = samples[i + 1].t - samples[i - 1].t;
    if (!(dt > 0)) continue;
    const trunk = trunkLengths[i];
    if (!(trunk > 1e-6)) continue;
    speed[i] = distance(relative[i + 1], relative[i - 1]) / dt / trunk;
  }
  if (speed.length > 2) {
    speed[0] = speed[1];
    speed[speed.length - 1] = speed[speed.length - 2];
  }
  return speed;
}

/**
 * Name the shot from the hand's position in the torso frame at contact.
 *
 * @returns {{shot: string, side: string, height: string, confidence: number}}
 */
export function classifyShot(hand, thresholds = SHOT_THRESHOLDS) {
  const { height, lateral } = hand;
  let heightClass;
  if (height >= thresholds.overheadHeight) heightClass = 'overhead';
  else if (height >= thresholds.underarmHeight) heightClass = 'drive';
  else heightClass = 'underarm';

  let sideClass;
  if (lateral >= thresholds.forehandLateral) sideClass = 'forehand';
  else if (lateral <= thresholds.backhandLateral) sideClass = 'backhand';
  else sideClass = 'roundhead';

  // A shot played over the head only makes sense as its own category when it is
  // actually overhead; lower down, the same lateral band is just a straight-on
  // forehand played close to the body.
  if (sideClass === 'roundhead' && heightClass !== 'overhead') sideClass = 'forehand';

  // How far the hand sits from the nearest decision boundary, as a rough
  // confidence. Shots near a boundary are the ones a human would also hesitate on.
  const sideMargin = Math.min(
    Math.abs(lateral - thresholds.forehandLateral),
    Math.abs(lateral - thresholds.backhandLateral),
  );
  const heightMargin = Math.min(
    Math.abs(height - thresholds.overheadHeight),
    Math.abs(height - thresholds.underarmHeight),
  );
  const confidence = Math.max(0, Math.min(1, (Math.min(sideMargin, heightMargin) / 0.25)));

  return {
    shot: `${sideClass}-${heightClass}`,
    side: sideClass,
    height: heightClass,
    confidence,
  };
}

function localPeaks(speed, options) {
  const peaks = [];
  const w = options.peakWindow;
  for (let i = w; i < speed.length - w; i += 1) {
    const v = speed[i];
    if (!(v >= options.peakSpeed)) continue;
    let dominant = true;
    for (let j = i - w; j <= i + w; j += 1) {
      if (j === i) continue;
      // `>` on the left and `>=` on the right keeps exactly one index on a plateau.
      if (j < i ? speed[j] >= v : speed[j] > v) {
        dominant = false;
        break;
      }
    }
    if (dominant) peaks.push(i);
  }
  return peaks;
}

function applyRefractory(peaks, speed, times, refractory) {
  // Strongest first, so that when two candidates are too close together we keep
  // the real contact rather than whichever came first in time.
  const byStrength = peaks.slice().sort((a, b) => speed[b] - speed[a]);
  const kept = [];
  for (const p of byStrength) {
    if (kept.every((k) => Math.abs(times[p] - times[k]) >= refractory)) kept.push(p);
  }
  return kept.sort((a, b) => a - b);
}

function walkTo(speed, times, from, direction, limitSeconds, quiet) {
  let i = from;
  while (true) {
    const next = i + direction;
    if (next < 0 || next >= speed.length) break;
    if (Math.abs(times[next] - times[from]) > limitSeconds) break;
    i = next;
    if (speed[i] <= quiet) break;
  }
  return i;
}

/**
 * Detect every stroke in a sequence of analysed frames.
 *
 * @param {Array} frames each `{ t, world, metrics, image? }`, world in the y-up
 *   frame and `metrics` from `biomech.frameMetrics`
 * @param {object} [options] see {@link DEFAULT_STROKE_OPTIONS}
 * @returns {Array} one record per stroke
 */
export function detectStrokes(frames, options = {}) {
  const opts = { ...DEFAULT_STROKE_OPTIONS, ...options };
  const thresholds = { ...SHOT_THRESHOLDS, ...(options.thresholds || {}) };
  if (frames.length < 5) return [];

  const times = frames.map((f) => f.t);
  const trunks = frames.map((f) => f.metrics.trunkLength);
  const racketArm = frames[0].metrics.racketArm;
  const speed = wristSpeedSeries(frames, racketArm, trunks);

  const peaks = applyRefractory(localPeaks(speed, opts), speed, times, opts.refractory);

  const strokes = [];
  for (const peak of peaks) {
    const visibility = frames[peak].image
      ? coreVisibility(frames[peak].image, CORE)
      : 1;
    if (visibility < opts.minVisibility) continue;

    const start = walkTo(speed, times, peak, -1, opts.maxBackswing, opts.quietSpeed);
    const end = walkTo(speed, times, peak, +1, opts.maxFollowThrough, opts.quietSpeed);
    const contact = frames[peak].metrics;
    const label = classifyShot(contact.hand, thresholds);

    const window = frames.slice(start, end + 1);
    const backswing = frames.slice(start, peak + 1);

    strokes.push({
      index: strokes.length,
      frame: frames[peak].frame,
      t: times[peak],
      startT: times[start],
      endT: times[end],
      startFrame: frames[start].frame,
      endFrame: frames[end].frame,
      ...label,
      visibility,
      peakSpeed: speed[peak],
      peakSpeedMs: speed[peak] * contact.trunkLength,
      backswingDuration: times[peak] - times[start],
      followThroughDuration: times[end] - times[peak],
      contact: {
        elbow: contact.elbow,
        elbowOff: contact.elbowOff,
        shoulderElevation: contact.shoulderElevation,
        shoulderElevationOff: contact.shoulderElevationOff,
        shoulderAzimuth: contact.shoulderAzimuth,
        separation: contact.separation,
        trunkLean: contact.trunkLean,
        kneeLeft: contact.kneeLeft,
        kneeRight: contact.kneeRight,
        stanceWidth: contact.stanceWidth,
        hand: {
          height: contact.hand.height,
          lateral: contact.hand.lateral,
          forward: contact.hand.forward,
          extension: contact.hand.extension,
        },
      },
      // A coach reads the backswing as much as the contact: the deepest elbow
      // bend and the largest hip-shoulder twist before contact are what the
      // arm had available to release.
      backswing: {
        minElbow: minBy(backswing, (f) => f.metrics.elbow),
        maxSeparation: maxBy(backswing, (f) => f.metrics.separation),
        maxShoulderElevation: maxBy(backswing, (f) => f.metrics.shoulderElevation),
        maxOffArmElevation: maxBy(backswing, (f) => f.metrics.shoulderElevationOff),
      },
      window: {
        maxElbow: maxBy(window, (f) => f.metrics.elbow),
        minKnee: minBy(window, (f) => Math.min(f.metrics.kneeLeft, f.metrics.kneeRight)),
      },
    });
  }
  return strokes;
}

function minBy(items, fn) {
  let best = Infinity;
  for (const item of items) {
    const v = fn(item);
    if (Number.isFinite(v) && v < best) best = v;
  }
  return Number.isFinite(best) ? best : NaN;
}

function maxBy(items, fn) {
  let best = -Infinity;
  for (const item of items) {
    const v = fn(item);
    if (Number.isFinite(v) && v > best) best = v;
  }
  return Number.isFinite(best) ? best : NaN;
}

/** Tally shots by name, for the session summary. */
export function summariseStrokes(strokes) {
  const byShot = {};
  for (const s of strokes) byShot[s.shot] = (byShot[s.shot] || 0) + 1;
  const speeds = strokes.map((s) => s.peakSpeedMs).filter(Number.isFinite);
  return {
    count: strokes.length,
    byShot,
    forehand: strokes.filter((s) => s.side === 'forehand').length,
    backhand: strokes.filter((s) => s.side === 'backhand').length,
    roundhead: strokes.filter((s) => s.side === 'roundhead').length,
    overhead: strokes.filter((s) => s.height === 'overhead').length,
    meanPeakSpeedMs: speeds.length ? speeds.reduce((a, b) => a + b, 0) / speeds.length : NaN,
    maxPeakSpeedMs: speeds.length ? Math.max(...speeds) : NaN,
  };
}
